api/layer/object.go

```diff
@@ -682,19 +699,19 @@ func IsSystemHeader(key string) bool {
 }
 
 func shouldSkip(node *data.NodeVersion, p allObjectParams, existed map[string]struct{}) bool {
-	filepath := node.FilePath
-	if len(p.Delimiter) > 0 {
-		tail := strings.TrimPrefix(filepath, p.Prefix)
-		index := strings.Index(tail, p.Delimiter)
-		if index >= 0 {
-			filepath = p.Prefix + tail[:index+1]
-		}
+	if node.IsDeleteMarker() {
+		return true
+	}
+
+	filePath := node.FilePath
+	if dirName := tryDirectoryName(node, p.Prefix, p.Delimiter); len(dirName) != 0 {
+		filePath = dirName
 	}
-	if _, ok := existed[filepath]; ok {
+	if _, ok := existed[filePath]; ok {
 		return true
```

Comment:
Could you please provide an example when we go to this line?
I coudn't find a suitable example.

Reply:
Let's say you have objects: `prefix/dir/obj1`, `prefix/dir/obj2`, `prefix/obj3` and you do listing with prefix: `prefix` and delimiter: `/`. In this case after you handle `prefix/dir/obj1` and get "directory" `prefix/dir/` you can skip handling all other objects with prefix `prefix/dir/` (in our case it's `prefix/dir/obj2`).
The final listing result will be:
```json
{
  "Content": [
    {
      "Key": "prefix/obj3"
    }
  ],
  "CommonPrefix": [
    {
      "Prefix": "prefix/dir/"
    }
  ]
}
```